File: core/src/progression.ts
// La progressione dei Rubyfront (2026-09-23): il gemello TypeScript di
// engine/lib/rubyfront/progression.rb. Un modulo a parte, fuori dalle
// regole: livelli dall'esperienza, abilità sbloccate per livello,
// configurazione (le abilità montate, a slot fissi per faccia) e la sua
// convalida — stessi motivi del tavolo, qui come chiavi i18n. I dati
// arrivano dal catalogo (`progression` in docs/cards/catalog.json), come
// i mazzi. Le abilità di oggi sono segnaposto: non si giocano ancora.

export type ProgressionFace = "rubyfront" | "nexus";
export const PROGRESSION_FACES: ProgressionFace[] = ["rubyfront", "nexus"];
export const PROGRESSION_LEVELS = 10;

export interface ProgressionRules {
  /** thresholds[i] è l'esperienza cumulata per essere al livello i+1. */
  thresholds: number[];
  points: { win: number; loss: number; draw: number };
  slots: Record<ProgressionFace, number>;
}

export interface ProgressionAbility {
  id: string;
  locales: Record<string, { name: string; text: string }>;
}

export interface ProgressionLevel {
  level: number;
  rubyfront: ProgressionAbility;
  nexus: ProgressionAbility;
}

export interface ProgressionCard {
  card: string;
  levels: ProgressionLevel[];
}

export interface ProgressionCatalog {
  rules: ProgressionRules;
  cards: ProgressionCard[];
}

/**
 * Le abilità montate, per faccia: la lista è POSIZIONALE (2026-09-24) —
 * l'indice è il blocco stampato sulla carta che l'abilità sostituisce, `null`
 * lascia quello stampato. Gemello di Progression#loadout_ok?.
 */
export type Loadout = Record<ProgressionFace, (string | null)[]>;

/** La progressione di un giocatore su un Rubyfront, come la manda il tavolo. */
export interface PlayerProgress {
  card: string;
  xp: number;
  level: number;
  loadout: Loadout;
}

export type Outcome = "win" | "loss" | "draw";

export const EMPTY_LOADOUT: Loadout = { rubyfront: [], nexus: [] };

let catalog: ProgressionCatalog | null = null;

/** Il catalogo della progressione, consegnato da `useCatalog` (cards.ts). */
export function useProgression(next: ProgressionCatalog | null | undefined): void {
  catalog = next ?? null;
}

export function progressionRules(): ProgressionRules | null {
  return catalog?.rules ?? null;
}

export function progressionOf(cardId: string): ProgressionCard | null {
  return catalog?.cards.find(card => card.card === cardId) ?? null;
}

/** I Rubyfront con una progressione, nell'ordine del catalogo. */
export function progressionCards(): ProgressionCard[] {
  return catalog?.cards ?? [];
}

/** Il livello dato l'esperienza: l'ultima soglia raggiunta, fra 1 e 10. Gemello: Progression#level_for. */
export function levelFor(xp: number, rules: ProgressionRules): number {
  const reached = rules.thresholds.filter(threshold => xp >= threshold).length;
  return Math.max(1, Math.min(PROGRESSION_LEVELS, reached));
}

/** Le soglie di partenza e d'arrivo di un livello (per la barra); l'ultimo non ha arrivo. Gemello: Progression#span. */
export function xpSpan(level: number, rules: ProgressionRules): { from: number; to: number | null } {
  const at = Math.max(1, Math.min(PROGRESSION_LEVELS, level));
  return { from: rules.thresholds[at - 1] ?? 0, to: at < PROGRESSION_LEVELS ? (rules.thresholds[at] ?? null) : null };
}

/** Gli id delle abilità sbloccate a quel livello, per faccia. Gemello: Progression#unlocked. */
export function unlocked(card: ProgressionCard | null, level: number): Loadout {
  const levels = (card?.levels ?? []).filter(entry => entry.level <= level);
  return { rubyfront: levels.map(entry => entry.rubyfront.id), nexus: levels.map(entry => entry.nexus.id) };
}

/** L'abilità con quell'id, con la sua faccia e il suo livello. */
export function abilityOf(card: ProgressionCard | null, id: string): { ability: ProgressionAbility; face: ProgressionFace; level: number } | null {
  for (const entry of card?.levels ?? []) {
    for (const face of PROGRESSION_FACES) if (entry[face].id === id) return { ability: entry[face], face, level: entry.level };
  }
  return null;
}

/** Il nome e il testo di un'abilità nella lingua data (la lingua del catalogo se manca). */
export function abilityWords(ability: ProgressionAbility, locale: string): { name: string; text: string } {
  return ability.locales[locale] ?? ability.locales.it ?? ability.locales.en ?? { name: ability.id, text: "" };
}

/**
 * La configurazione è lecita? `null` se sì, se no la chiave i18n del motivo
 * (con i suoi parametri). Gemello: Progression#loadout_ok? — stesse regole:
 * solo le due facce, id esistenti e sbloccati, niente doppioni, al più gli
 * slot della faccia.
 */
export function validateLoadout(card: ProgressionCard | null, level: number, loadout: Loadout, rules: ProgressionRules): { key: string; params?: Record<string, string | number> } | null {
  if (!card) return { key: "progression.refuse.unknown" };
  const open = unlocked(card, level);
  for (const face of PROGRESSION_FACES) {
    const slots = loadout[face] ?? [];
    if (slots.length > rules.slots[face]) return { key: "progression.refuse.slots", params: { face, n: rules.slots[face] } };
    const ids = slots.filter((id): id is string => typeof id === "string");
    if (new Set(ids).size !== ids.length) return { key: "progression.refuse.repeated", params: { face } };
    const missing = ids.find(id => !open[face].includes(id));
    if (missing) return { key: "progression.refuse.locked", params: { face, id: missing } };
  }
  return null;
}

/** La progressione «vuota» di un Rubyfront: livello 1, niente montato. */
export function freshProgress(cardId: string): PlayerProgress {
  return { card: cardId, xp: 0, level: 1, loadout: { rubyfront: [], nexus: [] } };
}

/** Le abilità montate di una faccia (senza i buchi). */
export function mountedIds(loadout: Loadout, face: ProgressionFace): string[] {
  return (loadout[face] ?? []).filter((id): id is string => typeof id === "string");
}

/** In quale blocco è montata l'abilità, o -1. */
export function slotOf(loadout: Loadout, face: ProgressionFace, id: string): number {
  return (loadout[face] ?? []).indexOf(id);
}

/** Una configurazione con quell'abilità montata nel blocco `slot` (tolta da dov'era), senza convalida: la convalida la fa chi decide. */
export function withAbility(loadout: Loadout, face: ProgressionFace, id: string, slot: number): Loadout {
  const next: Loadout = { rubyfront: [...loadout.rubyfront], nexus: [...loadout.nexus] };
  const list = next[face].map(other => (other === id ? null : other));
  while (list.length <= slot) list.push(null);
  list[slot] = id;
  next[face] = trimSlots(list);
  return next;
}

/** Una configurazione col blocco `slot` liberato (torna l'abilità stampata). */
export function withoutSlot(loadout: Loadout, face: ProgressionFace, slot: number): Loadout {
  const next: Loadout = { rubyfront: [...loadout.rubyfront], nexus: [...loadout.nexus] };
  const list = [...next[face]];
  if (slot < list.length) list[slot] = null;
  next[face] = trimSlots(list);
  return next;
}

/** I buchi in coda non contano: si tolgono. */
function trimSlots(list: (string | null)[]): (string | null)[] {
  const out = [...list];
  while (out.length && out[out.length - 1] === null) out.pop();
  return out;
}

File: engine/lib/rubyfront/progression.rb
# frozen_string_literal: true

require "json"

module Rubyfront
  # La progressione dei Rubyfront (2026-09-23): un modulo a parte, fuori dal
  # giudizio. Legge una volta `data/progression/` — le regole (soglie di
  # livello, punti per esito, slot per faccia) e un file per Rubyfront con i
  # dieci livelli, ciascuno con un'abilità Rubyfront e una Nexus — e congela.
  # Sa dire il livello dato l'esperienza, quanto vale una partita, quali
  # abilità sono sbloccate a un livello e se una configurazione (`loadout`)
  # è lecita. Non tocca la memoria né la rete: chi la usa (account, trasporto)
  # le passa i numeri e ne riceve i verdetti. Le abilità di oggi sono
  # segnaposto: l'engine non le legge ancora (lo farà dai tre punti
  # dichiarati nel README).
  class Progression
    FACES = %w[rubyfront nexus].freeze
    OUTCOMES = %w[win loss draw].freeze
    LEVELS = 10

    attr_reader :rules, :cards

    def self.load(data_dir)
      dir = File.join(data_dir, "progression")
      rules = JSON.parse(File.read(File.join(dir, "rules.json")))
      cards = Dir[File.join(dir, "*.json")].reject { |file| File.basename(file) == "rules.json" }.to_h do |file|
        doc = JSON.parse(File.read(file))
        [doc["card"], doc]
      end
      new(rules, cards)
    end

    def initialize(rules, cards)
      @rules = deep_freeze(rules)
      @cards = deep_freeze(cards)
      @thresholds = Array(@rules["thresholds"]).map(&:to_i).freeze
      @slots = FACES.to_h { |face| [face, @rules.dig("slots", face).to_i] }.freeze
      @points = OUTCOMES.to_h { |outcome| [outcome, @rules.dig("points", outcome).to_i] }.freeze
    end

    def known?(card_id)
      @cards.key?(card_id)
    end

    def card_ids
      @cards.keys.sort
    end

    # Il livello dato l'esperienza cumulata: l'ultima soglia raggiunta, mai
    # oltre il decimo livello, mai sotto il primo.
    def level_for(xp)
      xp = xp.to_i
      level = @thresholds.count { |threshold| xp >= threshold }
      level.clamp(1, LEVELS)
    end

    # Le soglie di partenza e d'arrivo di un livello (per la barra): l'ultimo
    # livello non ha arrivo.
    def span(level)
      level = level.to_i.clamp(1, LEVELS)
      [@thresholds[level - 1], level < LEVELS ? @thresholds[level] : nil]
    end

    def points_for(outcome)
      @points.fetch(outcome.to_s, 0)
    end

    def slots(face)
      @slots.fetch(face.to_s, 0)
    end

    # Le abilità sbloccate a quel livello, per faccia: gli id dei livelli
    # fino a `level` compreso.
    def unlocked(card_id, level)
      levels = Array(@cards.dig(card_id, "levels")).select { |entry| entry["level"].to_i <= level.to_i }
      FACES.to_h { |face| [face, levels.filter_map { |entry| entry.dig(face, "id") }] }
    end

    # La configurazione è lecita? Torna `[true]` oppure `[false, motivo,
    # motivo_en]`. Le chiavi sono solo le due facce; ogni lista è POSIZIONALE
    # (2026-09-24): l'indice è il blocco stampato sulla carta che l'abilità
    # sostituisce, `nil` lascia quello stampato; gli id devono esistere ed
    # essere sbloccati; niente doppioni; al più gli slot della faccia.
    def loadout_ok?(card_id, level, loadout)
      return [false, "Rubyfront sconosciuto alla progressione", "Rubyfront unknown to the progression"] unless known?(card_id)
      return [false, "la configurazione dev'essere un oggetto con rubyfront e nexus", "the loadout must be an object with rubyfront and nexus"] unless loadout.is_a?(Hash)

      extra = loadout.keys.map(&:to_s) - FACES
      return [false, "chiavi non previste: #{extra.join(", ")}", "unexpected keys: #{extra.join(", ")}"] unless extra.empty?

      open = unlocked(card_id, level)
      FACES.each do |face|
        slots_list = loadout[face] || loadout[face.to_sym] || []
        return [false, "#{face}: dev'essere una lista di id", "#{face}: must be a list of ids"] unless slots_list.is_a?(Array) && slots_list.all? { |id| id.nil? || id.is_a?(String) }
        return [false, "#{face}: al più #{slots(face)} abilità montate", "#{face}: at most #{slots(face)} mounted abilities"] if slots_list.size > slots(face)

        ids = slots_list.compact
        return [false, "#{face}: abilità ripetuta", "#{face}: repeated ability"] if ids.uniq.size != ids.size

        unknown = ids - open[face]
        return [false, "#{face}: abilità non sbloccata o sconosciuta (#{unknown.first})", "#{face}: ability locked or unknown (#{unknown.first})"] unless unknown.empty?
      end
      [true]
    end

    # La forma pubblica di una riga della memoria (o di nessuna riga: livello
    # 1, niente montato) per un Rubyfront.
    def progress_of(card_id, row = nil)
      xp = row ? row[:xp].to_i : 0
      loadout = row && row[:loadout].is_a?(Hash) ? row[:loadout] : {}
      {
        card: card_id,
        xp: xp,
        level: level_for(xp),
        loadout: FACES.to_h { |face| [face, Array(loadout[face] || loadout[face.to_sym]).map { |id| id.nil? ? nil : id.to_s }] }
      }
    end

    private

    def deep_freeze(value)
      case value
      when Hash then value.each_value { |inner| deep_freeze(inner) }.freeze
      when Array then value.each { |inner| deep_freeze(inner) }.freeze
      else value.freeze
      end
    end
  end
end
